### PR TITLE
Implement vcr.Verify

### LINKED ISSUE #117: VCR: Verify Credential
part of #66

When a VC is published over the network. It must be validated and we must make sure it's 100% correct.

When valid, it has to be added to the store.

### PR DESCRIPTION
**Pull request by woutslakhorst**
waiting for #100 
closes #117

### CODE REVIEW

File: vcr/test/vc.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
+{
+  "context": [
```
Comment (reinkrul) [inline vcr/test/vc.json:2]: @context?

File: vcr/vcr.go [hunk 2]
```diff
@@ -224,8 +227,62 @@ func (c *vcr) Resolve(ID string) (did.VerifiableCredential, error) {
 	panic("implement me")
 }
 
-func (c *vcr) Verify(vc did.VerifiableCredential, credentialSubject interface{}, at time.Time) (bool, error) {
-	panic("implement me")
+func (c *vcr) Verify(vc did.VerifiableCredential, at time.Time) error {
+	// it must have valid content
+	validator, _ := credential.FindValidatorAndBuilder(vc)
+	if validator == nil {
```
Comment (reinkrul) [inline vcr/vcr.go:233]: I think `Issue()` checks both validator and builder for `nil`? Let's decide on one style

File: vcr/vcr.go [hunk 3]
```diff
@@ -224,8 +227,62 @@ func (c *vcr) Resolve(ID string) (did.VerifiableCredential, error) {
...
+	// create correct challenge for verification
+	payload, err := generateCredentialChallenge(vc)
+	if err != nil {
+		return fmt.Errorf("cannot generate challenge: %w", err)
+	}
+
+	// extract proof, can't fail already done in generateCredentialChallenge
+	var proofs = make([]JSONWebSignature2020Proof, 0)
+	_ = vc.UnmarshalProofValue(&proofs)
+	proof := proofs[0]
```
Comment (reinkrul) [inline vcr/vcr.go:250]: Could there be multiple? In that case I'd say return error when there's multiple

File: vcr/vcr.go [hunk 4]
```diff
@@ -224,8 +227,62 @@ func (c *vcr) Resolve(ID string) (did.VerifiableCredential, error) {
...
+	}
+
+	// find key
+	pk, err := c.docResolver.ResolveSigningKey(proof.VerificationMethod.String(), &at)
+	if err != nil {
+		return err
+	}
+
+	// the proof must be correct
+	verifier, _ := jws.NewVerifier(jwa.ES256)
```
Comment (reinkrul) [inline vcr/vcr.go:267]: theoretically we support other keys in our DIDs than just RSA so should be check that?

File: vcr/vcr.go [hunk 5]
```diff
@@ -224,8 +227,62 @@ func (c *vcr) Resolve(ID string) (did.VerifiableCredential, error) {
...
+	// the jws lib can't do this for us, so we concat hdr with payload for verification
+	challenge := fmt.Sprintf("%s.%s", splittedJws[0], base64.RawURLEncoding.EncodeToString(payload))
+	if err = verifier.Verify([]byte(challenge), sig, pk); err != nil {
+		return err
+	}
+
+	// next check timeRestrictions
+	if vc.IssuanceDate.After(at) {
```
Comment (reinkrul) [inline vcr/vcr.go:275]: on or after, I'd think?

File: vcr/vcr.go [hunk 6]
```diff
@@ -224,8 +227,62 @@ func (c *vcr) Resolve(ID string) (did.VerifiableCredential, error) {
...
+		return errors.New("credential not valid yet at given time")
+	}
+	if vc.ExpirationDate != nil && (*vc.ExpirationDate).Before(at) {
+		return errors.New("credential not valid anymore at given time")
+	}
+
+	// check if issuer is trusted
+	// todo requires trusted config
```
Comment (reinkrul) [inline vcr/vcr.go:283]: add issue

**Review by reinkrul — CHANGES_REQUESTED** [review]
I'm missing "it is added to the store" part as stated by the related issue?

File: vcr/test/vc.json (hunk 1 above)
Comment (woutslakhorst) [reply vcr/test/vc.json:2]: fixed by future PR

File: vcr/vcr.go (hunk 2 above)
Comment (woutslakhorst) [reply vcr/vcr.go:233]: Only the validator is required here so no need to check the builder.

File: vcr/vcr.go (hunk 3 above)
Comment (woutslakhorst) [reply vcr/vcr.go:250]: Covered by the validator.

File: vcr/vcr.go (hunk 6 above)
Comment (woutslakhorst) [reply vcr/vcr.go:283]: already covered in #66

File: vcr/vcr.go (hunk 5 above)
Comment (woutslakhorst) [reply vcr/vcr.go:275]: I think it should be valid at the same nanosecond. So an inclusive start but exclusive end.

File: vcr/vcr.go (hunk 4 above)
Comment (woutslakhorst) [reply vcr/vcr.go]: added a `SigningAlgorithm` func to crypto/jwx to determine signing algo based on key.

**Comment by woutslakhorst** [comment]
There's a later PR that connects the VCR to the network. No worries

**Review by stevenvegt — APPROVED** [review]

**Review by reinkrul — APPROVED** [review]

**Pull request merged**